### ISSUE TITLE
Incorporating Levenshtein distance as another option in addition to Jaro-Winkler

### ISSUE THREAD

**Issue by Daniel-Liu-c0deb0t** [andylokandy/simsearch-rs#6]
Hey,
Thanks for your work on this library. I was wondering whether you would be interested in adding Levenshtein distance as an option? I am recently working on accelerating bounded Levenshtein distance calculations with SIMD, which is around 20-30x faster than a normal scalar Levenshtein distance algorithm for medium length strings, and I thought that it would be nice and beneficial to use it in your library. The library is on crates.io: https://crates.io/crates/triple_accel.

However, there are some tradeoffs though: my library cannot handle unicode strings and null bytes due to the SIMD implementation. This could be addressed by allowing the user to opt in to using the SIMD implementation explicitly. Of course, my library supports automatically falling back to a scalar version if SIMD is not supported as a CPU feature.

If you are interested, we could discuss a little more about the implementation. I am probably able to work on this in the future.

**Comment by andylokandy** [comment]
Thanks! I think it's nice to have. But I'm a bit confused by:

> my library cannot handle unicode strings and null bytes

Do you mean the library can only work on ASCII strings? If that is the case, it's indeed sorts of difficult...

**Comment by Daniel-Liu-c0deb0t** [comment]
Yes, it can only work on ASCII byte strings, because they are split into chunks of 16 or 32 bytes that are loaded into SSE or AVX SIMD vectors. This is a tradeoff, as it is more efficient, but also unable to handle unicode. Therefore, I was thinking that there should be a boolean option where the user can indicate what they want.

**Comment by lukevp** [comment]
@Daniel-Liu-c0deb0t , isn't UTF-8 using an 8 bit byte array representation of Unicode character space?  Do you mean only UTF-16 won't work, or is there something specific about 7-bit ASCII vs UTF-8 that makes UTF-8 not work?

**Comment by Daniel-Liu-c0deb0t** [comment]
UTF-8 is *variable-length*, meaning that each character can be anywhere from one byte to 4 bytes. My library assumes that the input is just a byte (`u8`) slice, so it will obviously incorrectly overestimate the number of character edits if the string is UTF-8.

**Comment by lukevp** [comment]
Got it, thanks!
